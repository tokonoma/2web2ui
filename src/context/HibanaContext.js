import React, { createContext, useContext } from 'react';
import { connect } from 'react-redux';
import { updateUserUIOptions } from 'src/actions/currentUser';
import { showAlert } from 'src/actions/globalAlert';
import { isUserUiOptionSet } from 'src/helpers/conditions/user';
import { isAccountUiOptionSet } from 'src/helpers/conditions/account';
import { selectCondition } from 'src/selectors/accessConditionState';

const HibanaStateContext = createContext();

function Provider(props) {
  const { isHibanaEnabled: _isHibanaEnabled, children, ...rest } = props;

  // TODO: When OG theme is removed, the env variables will no longer be necessary
  const isHibanaEnabled =
    process.env.REACT_APP_DEFAULT_TO_HIBANA === 'true' ? true : _isHibanaEnabled;

  return (
    <HibanaStateContext.Provider value={{ isHibanaEnabled, ...rest }}>
      {children}
    </HibanaStateContext.Provider>
  );
}

function mapStateToProps(state) {
  return {
    isHibanaEnabled: selectCondition(isUserUiOptionSet('isHibanaEnabled'))(state),
    isBannerVisible: selectCondition(isUserUiOptionSet('isHibanaBannerVisible'))(state),
    hasThemeControls: selectCondition(isAccountUiOptionSet('hibana.hasThemeControls'))(state),
  };
}

const mapDispatchToProps = {
  dismissBanner: () => updateUserUIOptions({ isHibanaBannerVisible: false }),
  setIsHibanaEnabled: bool => {
    if (window.pendo && window.pendo.track) {
      window.pendo.track(`Hibana Toggle - ${Boolean(bool) ? 'On' : 'Off'}`);
    }

    // Always dismiss the banner when the user toggles their theme
    return updateUserUIOptions({ isHibanaEnabled: bool, isHibanaBannerVisible: false });
  },
  showAlert,
};

export function useHibana() {
  const context = useContext(HibanaStateContext);

  if (context === undefined) throw new Error('useHibana must be used within a HibanaProvider');

  return [context];
}

export const HibanaProvider = connect(mapStateToProps, mapDispatchToProps)(Provider);

export function HibanaConsumer({ children }) {
  return <HibanaStateContext.Consumer>{children}</HibanaStateContext.Consumer>;
}
