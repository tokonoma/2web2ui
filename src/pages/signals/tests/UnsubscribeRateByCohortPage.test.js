import React from 'react';
import { shallow } from 'enzyme';
import { useHibana } from 'src/context/HibanaContext';
import { UnsubscribeRateByCohortPage } from '../UnsubscribeRateByCohortPage';

jest.mock('src/context/HibanaContext');

describe('Signals Unsubscribe Rate Page', () => {
  const data = [
    {
      date: '2017-01-01',
      c_total: 1,
    },
    {
      date: '2017-01-02',
      c_total: 10,
    },
  ];

  const dataEngRecency = [
    {
      date: '2017-01-01',
      c_uneng: 0.25,
    },
    {
      date: '2017-01-02',
      c_uneng: 0.5,
    },
  ];

  const defaultProps = {
    facetId: 'test.com',
    facet: 'sending-domain',
    handleDateSelect: jest.fn(),
    gap: 0.25,
    loading: false,
    empty: false,
    xTicks: [1, 2],
    selectedDate: '2017-01-02',
    shouldHighlightSelected: false,
    data,
    dataEngRecency,
  };

  const subject = props => {
    useHibana.mockImplementationOnce(() => [{ isHibanaEnabled: false }]);

    return shallow(<UnsubscribeRateByCohortPage {...defaultProps} {...props} />);
  };

  it('renders correctly', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    const wrapper = subject({ loading: true });

    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('does not render SpamTrapsPreview when facet is mb_provider', () => {
    const wrapper = subject({ facet: 'mb_provider' });

    expect(
      wrapper.find('withRouter(Connect(WithDetails(SpamTrapsPreview)))'),
    ).not.toContainMatchingElement();
  });

  it('renders empty correctly', () => {
    const wrapper = subject({ empty: true });

    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders error correctly', () => {
    const wrapper = subject({ error: { message: 'error message' } });

    expect(wrapper.find('Callout').prop('children')).toEqual('error message');
    expect(wrapper.find('Callout').prop('title')).toEqual('Unable to Load Data');
  });

  describe('bar chart props', () => {
    it('renders tooltip content', () => {
      const wrapper = subject();
      const Tooltip = wrapper.find('LineChart').prop('tooltipContent');

      expect(
        shallow(
          <Tooltip
            payload={{
              p_uneng_unsub: 0.11111,
              p_365d_unsub: 0.2,
              p_90d_unsub: 0.3,
              p_14d_unsub: 0.4,
              p_new_unsub: 0.5,
              date: '2018-01-01',
              p_total_unsub: 10,
            }}
          />,
        ),
      ).toMatchSnapshot();
    });

    it('gets x axis props', () => {
      const wrapper = subject();
      const axisProps = wrapper.find('LineChart').prop('xAxisProps');

      expect(axisProps.ticks).toEqual([1, 2]);
      expect(axisProps.tickFormatter('2018-12-05')).toEqual('12/5');
    });

    it('gets y axis props with default domain', () => {
      const wrapper = subject({ data: [{ p_total_unsub: 0 }, { p_total_unsub: null }] });
      const axisProps = wrapper.find('LineChart').prop('yAxisProps');

      expect(axisProps.tickFormatter(0.252344)).toEqual('25.234%');
      expect(axisProps.domain).toEqual([0, 1]);
    });

    it('gets y axis props with domain', () => {
      const wrapper = subject({ data: [{ p_total_unsub: 0.5 }, { p_total_unsub: 0.6 }] });
      const axisProps = wrapper.find('LineChart').prop('yAxisProps');

      expect(axisProps.domain).toEqual(['auto', 'auto']);
    });
  });
});
