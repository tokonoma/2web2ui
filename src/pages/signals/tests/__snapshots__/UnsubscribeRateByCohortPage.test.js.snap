// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals Unsubscribe Rate Page bar chart props renders tooltip content 1`] = `
<Fragment>
  <TooltipMetric
    color="#88E2E7"
    description="Never engaged, first email in last 7 days"
    key="new"
    label="New"
    value="50%"
  />
  <TooltipMetric
    color="#1A838B"
    description="0-14 days since last engagement"
    key="14d"
    label="Recently Engaged"
    value="40%"
  />
  <TooltipMetric
    color="#219EA8"
    description="14-90 days since last engagement"
    key="90d"
    label="Semi Recently Engaged"
    value="30%"
  />
  <TooltipMetric
    color="#29B9C7"
    description="90-365 days since last engagement"
    key="365d"
    label="Not Recently Engaged"
    value="20%"
  />
  <TooltipMetric
    color="#50CFDA"
    description="365+ days since last engagement"
    key="uneng"
    label="Never Engaged"
    value="11.111%"
  />
</Fragment>
`;

exports[`Signals Unsubscribe Rate Page renders correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Engagement Recency Overview",
      "to": "/signals/engagement",
    }
  }
  facet="sending-domain"
  facetId="test.com"
  title="Unsubscribe Rate by Cohort"
>
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        data-id="unsubscribe-rate-chart"
      >
        <withRouter(Tabs)
          facet="sending-domain"
          facetId="test.com"
        />
        <Panel.Section>
          <div
            className="LiftTooltip"
          >
            <LineChart
              height={300}
              lineType="linear"
              lines={
                Array [
                  Object {
                    "c_total": 1,
                    "date": "2017-01-01",
                  },
                  Object {
                    "c_total": 10,
                    "date": "2017-01-02",
                  },
                ]
              }
              margin={
                Object {
                  "bottom": 5,
                  "left": 18,
                  "right": 0,
                  "top": 12,
                }
              }
              onClick={[MockFunction]}
              selected="2017-01-02"
              shouldHighlightSelected={false}
              stroke="#000"
              tooltipContent={[Function]}
              tooltipWidth="250px"
              width="99%"
              xAxisProps={
                Object {
                  "tickFormatter": [Function],
                  "ticks": Array [
                    1,
                    2,
                  ],
                }
              }
              xKey="date"
              yAxisProps={
                Object {
                  "domain": Array [
                    0,
                    1,
                  ],
                  "tickFormatter": [Function],
                }
              }
              yKey="value"
              yKeys={
                Array [
                  Object {
                    "OGFill": "#1A838B",
                    "OGStroke": "#1A838B",
                    "description": "0-14 days since last engagement",
                    "fill": "#1A838B",
                    "hibanaFill": "#1a497d",
                    "hibanaStroke": "#1a497d",
                    "key": "p_14d_unsub",
                    "label": "Recently Engaged",
                    "stroke": "#1A838B",
                  },
                  Object {
                    "OGFill": "#219EA8",
                    "OGStroke": "#219EA8",
                    "description": "14-90 days since last engagement",
                    "fill": "#219EA8",
                    "hibanaFill": "#0060cc",
                    "hibanaStroke": "#0060cc",
                    "key": "p_90d_unsub",
                    "label": "Semi Recently Engaged",
                    "stroke": "#219EA8",
                  },
                  Object {
                    "OGFill": "#29B9C7",
                    "OGStroke": "#29B9C7",
                    "description": "90-365 days since last engagement",
                    "fill": "#29B9C7",
                    "hibanaFill": "#1273e6",
                    "hibanaStroke": "#1273e6",
                    "key": "p_365d_unsub",
                    "label": "Not Recently Engaged",
                    "stroke": "#29B9C7",
                  },
                  Object {
                    "OGFill": "#50CFDA",
                    "OGStroke": "#50CFDA",
                    "description": "365+ days since last engagement",
                    "fill": "#50CFDA",
                    "hibanaFill": "#429bff",
                    "hibanaStroke": "#429bff",
                    "key": "p_uneng_unsub",
                    "label": "Never Engaged",
                    "stroke": "#50CFDA",
                  },
                  Object {
                    "OGFill": "#88E2E7",
                    "OGStroke": "#88E2E7",
                    "description": "Never engaged, first email in last 7 days",
                    "fill": "#88E2E7",
                    "hibanaFill": "#78b6ff",
                    "hibanaStroke": "#78b6ff",
                    "key": "p_new_unsub",
                    "label": "New",
                    "stroke": "#88E2E7",
                  },
                ]
              }
              yRange={
                Array [
                  "auto",
                  "auto",
                ]
              }
            />
            <Legend
              items={
                Array [
                  Object {
                    "OGFill": "#88E2E7",
                    "OGStroke": "#88E2E7",
                    "description": "Never engaged, first email in last 7 days",
                    "hibanaFill": "#78b6ff",
                    "hibanaStroke": "#78b6ff",
                    "label": "New",
                  },
                  Object {
                    "OGFill": "#50CFDA",
                    "OGStroke": "#50CFDA",
                    "description": "365+ days since last engagement",
                    "hibanaFill": "#429bff",
                    "hibanaStroke": "#429bff",
                    "label": "Never Engaged",
                  },
                  Object {
                    "OGFill": "#29B9C7",
                    "OGStroke": "#29B9C7",
                    "description": "90-365 days since last engagement",
                    "hibanaFill": "#1273e6",
                    "hibanaStroke": "#1273e6",
                    "label": "Not Recently Engaged",
                  },
                  Object {
                    "OGFill": "#219EA8",
                    "OGStroke": "#219EA8",
                    "description": "14-90 days since last engagement",
                    "hibanaFill": "#0060cc",
                    "hibanaStroke": "#0060cc",
                    "label": "Semi Recently Engaged",
                  },
                  Object {
                    "OGFill": "#1A838B",
                    "OGStroke": "#1A838B",
                    "description": "0-14 days since last engagement",
                    "hibanaFill": "#1a497d",
                    "hibanaStroke": "#1a497d",
                    "label": "Recently Engaged",
                  },
                ]
              }
              tooltipContent={[Function]}
            />
          </div>
        </Panel.Section>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <div>
        <Box
          as={[Function]}
          sectioned={true}
        >
          <UnsubscribeRateByCohortActions
            date="2017-01-02"
            recencyByCohort={
              Object {
                "c_uneng": 0.5,
                "date": "2017-01-02",
              }
            }
            unsubscribeByCohort={
              Object {
                "c_total": 10,
                "date": "2017-01-02",
              }
            }
          />
        </Box>
      </div>
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Unsubscribe Rate Page renders empty correctly 1`] = `
Array [
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>,
  <Panel
    data-id="unsubscribe-rate-chart"
  >
    <withRouter(Tabs)
      facet="sending-domain"
      facetId="test.com"
    />
    <Panel.Section>
      <Callout
        title="No Data Available"
      >
        Insufficient data to populate this chart
      </Callout>
    </Panel.Section>
  </Panel>,
]
`;

exports[`Signals Unsubscribe Rate Page renders loading correctly 1`] = `
Array [
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>,
  <Panel
    data-id="unsubscribe-rate-chart"
  >
    <withRouter(Tabs)
      facet="sending-domain"
      facetId="test.com"
    />
    <Panel.Section>
      <div
        style={
          Object {
            "height": "220px",
            "position": "relative",
          }
        }
      >
        <Loading />
      </div>
    </Panel.Section>
  </Panel>,
]
`;
